### DISCUSSION TITLE
How handle uniform to keep state during a draw call ?

### DISCUSSION THREAD

**Discussion by ra1jin** [gpuweb/gpuweb#3259] [Q&A] [answered]
Hello, I don't know why but when I make a succession of calls draw functions a strange thing appears.
It's as if the shader's uniform matrix takes the last value I set to it and ignores the others.
Is there a way to make the shader to take that matrix for that specific draw call and keep when shader work ?

I imagine that all this is due to the fact that the work only starts when the list of commands is sent to the device (at the end) and that is why the uniform matrix is ​​the last one I defined.

```
drawMesh(modelMatrix, vertexCount, vertexBuffer, texture) {
    this.passEncoder.setPipeline(this.triangleShaderRes.pipeline);
    let mvpcMatrix = new Float32Array(Utils.MAT4_MULTIPLY(this.vpcMatrix, modelMatrix));
    this.device.queue.writeBuffer(this.triangleShaderRes.uniformMVCPCMatrixBuffer, 0, mvpcMatrix.buffer);
    this.passEncoder.setBindGroup(0, this.triangleShaderRes.uniformMVCPCMatrixGroup);
    this.passEncoder.setBindGroup(1, texture ? texture.group : this.defaultTexture.group);
    this.passEncoder.setVertexBuffer(0, vertexBuffer);
    this.passEncoder.draw(vertexCount);
  }

```

PS: Sorry for the stupid question, i think i just need to create one matrix buffer by object to keep all alive... right ?

**Comment by Kangz** [comment, marked as answer]
The problem here is that there are two different timelines:

 - `writeBuffer` happens "immediately"
 - all other commands only happen when you submit them

So in the end what happens on the GPU looks like:

 - `writeBuffer` for the matrix of mesh1
 - `writeBuffer` for the matrix of mesh2
 - draw mesh1
 - draw mesh2

So only the matrix for mesh2 gets used.

What you should do is create a uniform buffer with enough space for all the matrices for the meshes in your frame. Then `writeBuffer` each matrix at a different offset, and `setBindGroup` such that each mesh uses the uniform buffer at a different offset. (you can use [dynamic offsets](https://gpuweb.github.io/gpuweb/#dom-gpubindingcommandsmixin-setbindgroup-index-bindgroup-dynamicoffsetsdata-dynamicoffsetsdatastart-dynamicoffsetsdatalength) for that).

**Reply by ra1jin** [reply]
Thank you for your quick and efficient response once again.
I understand your solution but it requires knowing the number of objects I'm going to draw and that's not the case for me.
I know you say that because binding groups are the fastest way to update information, but in my case this complicate a lot my renderer. Create an array for each object matrix is really penalizing for performance ?

**Reply by Kangz** [reply]
Allocating one buffer per matrix is going to be pretty inefficient. However what you could do is have large uniform buffers (say 10000 matrices worth) and allocate other ones if needed. This way you don't need to know the number of meshes in advance but still don't end up with too many small buffers.

**Reply by ra1jin** [reply]
So I must prefer always large buffer to small ones !
And to be complete on this subject, what about one large vertex buffer for all drawable ? Maybe too excessive ?
This is my last question :)

**Reply by Kangz** [reply]
In general having a few large buffers is better than having many small ones. That applies in particular to vertex buffers, uniform buffers etc. However sometimes having small buffers is better for structuring the application. So it's a tradeoff that you need to make, knowing the general guidance above.

**Reply by ra1jin** [reply]
I just done the batching for matrix and i do same thing for mesh vertex for a better consistence.
I post the code here for a little code review about optimization and to be sure i'm not doing mistake.
What about the cost of destroy/createBuffer and createBindGroup ?
My rendering code here :
```
beginDrawing(viewIndex) {
    // clear viewport, etc...

    this.meshDrawCount = 0;
    this.meshVertexCount = 0;
    this.meshCommands = [];

    this.debugDrawCount = 0;
    this.debugVertexCount = 0;
    this.debugCommands = [];
}

drawMesh(modelMatrix, vertexCount, vertices, texture) {
    let cmd = [];
    cmd[CMD_MATRIX_BUFFER_DATA] = Utils.MAT4_MULTIPLY(this.vpcMatrix, modelMatrix);
    cmd[CMD_MATRIX_BUFFER_OFFSET] = this.meshDrawCount * 16 * 4;
    cmd[CMD_VERTEX_BUFFER_DATA] = vertices;
    cmd[CMD_VERTEX_BUFFER_OFFSET] = this.meshVertexCount * 5 * 4;
    cmd[CMD_VERTEX_BUFFER_SIZE] = vertexCount * 5 * 4;
    cmd[CMD_VERTEX_COUNT] = vertexCount;
    cmd[CMD_TEXTURE_GROUP] = texture ? texture.group : this.defaultTexture.group;
    this.meshVertexCount += vertexCount;
    this.meshDrawCount++;
    this.meshCommands.push(cmd);
}

endDrawing() {
    // mesh shader
    // ------------------------------------------------------------------------------------
    this.passEncoder.setPipeline(this.meshPipeline);

    this.meshVertexBuffer.destroy();
    this.meshVertexBuffer = this.device.createBuffer({
      size: this.meshVertexCount * 5 * 4,
      usage: GPUBufferUsage.VERTEX | GPUBufferUsage.COPY_DST
    });

    this.meshMatrixBuffer.destroy();
    this.meshMatrixBuffer = this.device.createBuffer({
      size: this.meshDrawCount * 16 * 4,
      usage: GPUBufferUsage.UNIFORM | GPUBufferUsage.COPY_DST
    });

    for (let cmd of this.meshCommands) {
      let meshMatrixBinding = this.device.createBindGroup({
        layout: this.meshPipeline.getBindGroupLayout(0),
        entries: [{
          binding: 0,
          resource: {
            buffer: this.meshMatrixBuffer,
            offset: cmd[CMD_MATRIX_BUFFER_OFFSET],
            size: this.meshDrawCount * 16 * 4
          }
        }]
      });

      this.device.queue.writeBuffer(this.meshVertexBuffer, cmd[CMD_VERTEX_BUFFER_OFFSET], new Float32Array(cmd[CMD_VERTEX_BUFFER_DATA]));
      this.device.queue.writeBuffer(this.meshMatrixBuffer, cmd[CMD_MATRIX_BUFFER_OFFSET], new Float32Array(cmd[CMD_MATRIX_BUFFER_DATA]));
      this.passEncoder.setBindGroup(0, meshMatrixBinding);
      this.passEncoder.setBindGroup(1, cmd[CMD_TEXTURE_GROUP]);
      this.passEncoder.setVertexBuffer(0, this.meshVertexBuffer, cmd[CMD_VERTEX_BUFFER_OFFSET], cmd[CMD_VERTEX_BUFFER_SIZE]);
      this.passEncoder.draw(cmd[CMD_VERTEX_COUNT]);
}
```
